### ISSUE TITLE
Intellisense speedup proposal

### ISSUE THREAD

**Issue by piotr-rarus** [Microsoft/vscode#61574]
Dear VSCode developers, contributors, users

I fell in love with VSCode, but recent Intellisense performance drop, puts my feeling to test.
I don't know much about Intellisense implementation, but my guess is, that it does lazy evaluation of
docstrings in imported modules, when displaying available prompts. Certainly this improves
memory consumption. Personally, I use VSCode for ML/CV related projects in Python.
Python is by far, most reasonable tool for such field, because there're a lot of great libs.
Downside is that my whole lib stack (NumPy, SciPy, Pandas, Scikit-Learn, Scikit-Image,, OpenCV, Imbalanced-Learn, Matplotlib, Seaborn, Keras, Tensorflow, PyTorch, etc, and much more of their dependencies), is quite heavy. Don't know much about other languages/implementations, but in CPython, it takes few seconds to startup a machine and load up dependencies before execution.

As developers, we mostly work on decent machines (16GB at least, I think 32/64GB is standard for ML/CV applications). For me, it wouldn't be much of expense (just a few GB of RAM) to load doc-strings from all project dependencies eagerly, and pre-cache them. Certainly, startup will take few minutes, but I can use that time to make a coffee (which I eventually always do, when I start my work). I'd rather waste few minutes for startup (which is one-time expense), and rejoice later upon Intellisense great performance, than pulling my hair out, during whole day of work.

Kind Regards,
Piotr Rarus

**Comment by weinand** [comment]
For what language are you seeing a "Intellisense performance drop"?
For Python?

**Comment by piotr-rarus** [comment]
Yes, I work mostly in Python, though I heard from my colleagues, which work with JS, front-end stack, that they also observe significant drop. JS is also interpreted language.

To be more exact: we're observing 15-30 second lag, which makes Intellisense practically unusable.

**Comment by piotr-rarus** [comment]
Another proposal:
to keep startup time in place, we could only pre-cache docstrings from opened files.
Import statements could be traversed one by one, and docstrings collected.
Imported dependencies should be observed to re-evaluate docstrings cache.
This would keep RAM usage in place too.

Final performance would be also CPU/RAM efficient, as docstring should be kept in dictionary-like collection. Dictionaries are sweet and fast, and we would only keep docstrings as values (strings aren't that much heavy). When prompting Intellisense, it would fetch docstring from cache and render it.
Accessing values in dictionary is super-super-super-fast, and rendering wouldn't also be a problem.

**Comment by piotr-rarus** [comment]
PS

Cache should be shared among opened files/modules.

**Comment by piotr-rarus** [comment]
To make it even more efficient:
when considering interpreted languages i.e. Python, JS, etc., we don't even need to import those dependencies. They're not compiled/pre-compiled. We could just as easily load them as text files, and fetch docstrings using simple regex.

**Comment by piotr-rarus** [comment]
PS3
Tracking dependencies wouldn't also be much of a problem. Could also be done using regex and simple parsing.

**Comment by jrieken** [comment]
> I fell in love with VSCode, but recent Intellisense performance drop, puts my feeling to test.

Can you describe in *detail* what you mean? Does it take longer to show the IntelliSense window (e.g. you see `Loading...` for long) or does it take longer to show the extra information (doc string) that appears when focusing an item?

Generally, this is nothing we compute - vscode is just fancy UI and data provider APIs. The actual data (completions, doc comments, etc) are all computed by extensions running in a separate process and we rely on them being fast. 

So, in what version do you see this slow/fast? Does this happens when *only* that extension is enabled (disable all others). Last, please profile the extension host when this happens, "F1 > Show Running Extensions" and then hit the circle button.

**Comment by piotr-rarus** [comment]
It displays little window, that just shows 'Loading...'. Rendering of fetched docstring lag is unnoticeable.

Could you please tell me, where profiler telemetry is dumped?

**Comment by piotr-rarus** [comment]
I apologize for writing this issue here. It should be moved to vscode-python repository.

**Comment by piotr-rarus** [comment]
I think it would be nice, to provide some guidelines and tips for extension's developers, when it comes to Intellisense performance. I know some of this knowledge is proprietary, but when it comes to Python extension in particular, well, Microsoft owns this. It's just kinda weird that Intellisense works marvelous in VSStudio, but lags in Python extension.
